palace/fem/libceed/coefficient.hpp

```diff
@@ -0,0 +1,131 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef PALACE_LIBCEED_COEFFICIENT_HPP
+#define PALACE_LIBCEED_COEFFICIENT_HPP
+
+#include <vector>
+#include <mfem.hpp>
+#include <mfem/linalg/dtensor.hpp>
+
+namespace palace::ceed
+{
+
+struct QuadratureCoefficient
+{
+  int ncomp;
+  mfem::Vector data;
+};
+
+inline void InitCoefficient(mfem::Coefficient &Q, mfem::ParMesh &mesh,
+                            const mfem::IntegrationRule &ir,
+                            const std::vector<int> &indices, bool use_bdr,
+                            QuadratureCoefficient &coeff)
+{
+  const auto ne = indices.size();
+  const auto nqpts = ir.GetNPoints();
+  coeff.ncomp = 1;
+  coeff.data.SetSize(ne * nqpts);
+  auto C = mfem::Reshape(coeff.data.HostWrite(), nqpts, ne);
+  mfem::IsoparametricTransformation T;
+  for (std::size_t i = 0; i < ne; ++i)
+  {
+    const auto e = indices[i];
+    if (use_bdr)
+    {
+      mesh.GetBdrElementTransformation(e, &T);
+    }
+    else
+    {
+      mesh.GetElementTransformation(e, &T);
+    }
+    for (int q = 0; q < nqpts; ++q)
+    {
+      const mfem::IntegrationPoint &ip = ir.IntPoint(q);
+      T.SetIntPoint(&ip);
+      C(q, i) = Q.Eval(T, ip);
+    }
+  }
+}
+
+inline void InitCoefficient(mfem::VectorCoefficient &VQ, mfem::ParMesh &mesh,
+                            const mfem::IntegrationRule &ir,
+                            const std::vector<int> &indices, bool use_bdr,
+                            QuadratureCoefficient &coeff)
+{
+  const auto ne = indices.size();
+  const auto vdim = VQ.GetVDim();
+  const auto nqpts = ir.GetNPoints();
+  coeff.ncomp = vdim;
+  coeff.data.SetSize(ne * nqpts * vdim);
+  auto C = mfem::Reshape(coeff.data.HostWrite(), vdim, nqpts, ne);
+  mfem::IsoparametricTransformation T;
+  mfem::DenseMatrix Q_ip(vdim, nqpts);
+  for (std::size_t i = 0; i < ne; ++i)
+  {
+    const auto e = indices[i];
+    if (use_bdr)
+    {
+      mesh.GetBdrElementTransformation(e, &T);
+    }
+    else
+    {
+      mesh.GetElementTransformation(e, &T);
+    }
+    VQ.Eval(Q_ip, T, ir);
+    for (int q = 0; q < nqpts; ++q)
+    {
+      for (int d = 0; d < vdim; ++d)
+      {
+        C(d, q, i) = Q_ip(d, q);
+      }
+    }
+  }
+}
+
+inline void InitCoefficient(mfem::MatrixCoefficient &MQ, mfem::ParMesh &mesh,
+                            const mfem::IntegrationRule &ir,
+                            const std::vector<int> &indices, bool use_bdr,
+                            QuadratureCoefficient &coeff)
+{
+  // Assumes matrix coefficient is symmetric.
```

Comment:
Worth having an `MFEM_ASSERT` for this?

Reply:
Supporting non-symmetric operators would require a lot more extensive changes throughout the code where this assumption is made so it's fine here too I would say.